instrumentation/mongo/mongo-3.1/library/src/main/java/io/opentelemetry/instrumentation/mongo/v3_1/TracingCommandListener.java

```diff
@@ -10,38 +10,47 @@
 import com.mongodb.event.CommandStartedEvent;
 import com.mongodb.event.CommandSucceededEvent;
 import io.opentelemetry.context.Context;
+import io.opentelemetry.instrumentation.api.instrumenter.Instrumenter;
 import java.util.Map;
 import java.util.concurrent.ConcurrentHashMap;
 
 final class TracingCommandListener implements CommandListener {
 
-  private final MongoClientTracer tracer;
+  private final Instrumenter<CommandStartedEvent, Void> instrumenter;
   private final Map<Integer, Context> contextMap;
+  private final Map<Integer, CommandStartedEvent> requestMap;
 
-  TracingCommandListener(MongoClientTracer tracer) {
-    this.tracer = tracer;
-    contextMap = new ConcurrentHashMap<>();
+  TracingCommandListener(Instrumenter<CommandStartedEvent, Void> instrumenter) {
+    this.instrumenter = instrumenter;
+    this.contextMap = new ConcurrentHashMap<>();
+    this.requestMap = new ConcurrentHashMap<>();
   }
 
   @Override
   public void commandStarted(CommandStartedEvent event) {
-    Context context = tracer.startSpan(Context.current(), event, event.getCommand());
-    contextMap.put(event.getRequestId(), context);
+    Context parentContext = Context.current();
+    if (instrumenter.shouldStart(parentContext, event)) {
+      Context context = instrumenter.start(parentContext, event);
+      contextMap.put(event.getRequestId(), context);
+      requestMap.put(event.getRequestId(), event);
+    }
   }
 
   @Override
   public void commandSucceeded(CommandSucceededEvent event) {
     Context context = contextMap.remove(event.getRequestId());
-    if (context != null) {
-      tracer.end(context);
+    CommandStartedEvent request = requestMap.get(event.getRequestId());
```

Comment:
Shouldn't it be 
```suggestion
    CommandStartedEvent request = requestMap.remove(event.getRequestId());
```
?

Also, WDYT about maintaining a single map containing a `ContextAndRequest` values?

Reply:
@iNikem this seems like a good idea